### ISSUE TITLE
JPA entities containing LAZY loaded relations cannot be merged in another transaction

### ISSUE THREAD

**Issue by cghislai** [quarkusio/quarkus#38807]
### Describe the bug

If I have  3 entities A,B,C;  a relation in B `@JoinColumn(fetch = LAZY) A a` and a relation in C `@JoinColumn B b`; then

If I fetch B in a JTA tranacstion, then merge it (calling EntityManager::merge) in another transaction, I get an error:
```
ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (executor-thread-1) HTTP Request to /hello/test failed, error id: 9ddaf11b-22a6-4e89-b324-733ef2c96b6f-1: org.hibernate.PropertyValueException: Detached entity with generated id '1' has an uninitialized version value 'null' : org.acme.entities.EntityA.version
	at org.hibernate.persister.entity.AbstractEntityPersister.isTransient(AbstractEntityPersister.java:4168)
	at org.hibernate.engine.internal.ForeignKeys.isTransient(ForeignKeys.java:293)
	at org.hibernate.type.EntityType.replace(EntityType.java:315)
	at org.hibernate.type.TypeHelper.replace(TypeHelper.java:88)
	at org.hibernate.event.internal.DefaultMergeEventListener.copyValues(DefaultMergeEventListener.java:576)
	at org.hibernate.event.internal.DefaultMergeEventListener.entityIsDetached(DefaultMergeEventListener.java:427)
	at org.hibernate.event.internal.DefaultMergeEventListener.merge(DefaultMergeEventListener.java:208)

```
If I fetch C or A in a JTA transaction, then merge them in another one, it passes.

EDIT: The issue only occurs when the entity behind the lazy relation has a `@Version` field. In the reproducer, that would be `EntityA#version`.

The JPA spec states at https://jakarta.ee/specifications/persistence/3.0/jakarta-persistence-spec-3.0#merging-detached-entity-state:

> The persistence provider must not merge fields marked LAZY that have not been fetched: it must ignore such fields when merging.

However here it seems quarkus is not ignoring the `a` field in `A` that is marked LAZY; and instead throws because the proxy does not contain the version value.

EDIT: spec has a paragraph about version field in that section as well. However, if I understand correctly it applies to the detached entity being merged, not the one behind the lazy relation.

> Any Version columns used by the entity must be checked by the persistence runtime implementation during the merge operation and/or at flush or commit time. In the absence of Version columns there is no additional version checking done by the persistence provider runtime during the merge operation.

But If I merge a parent entity, like C, then it passes, and I can then merge B.

### Expected behavior

I can fetch and merge JPA entities across tranasctions regarless of the columns FETCH type

### Actual behavior

When I have an entity with a column maked with fetch type LAZY; I cannot merge it directly.

I would need to merge a parent entity to obtain a managed instance.

### How to Reproduce?

The https://github.com/cghislai/quarkus-merge-lazy contains a reproducer test.

- Notice how, with line 37 commented, I can merge entityB through its parent entityC.
- Uncomment line 37 in org.acme.GreetingResource#hello to reproduce the error

**Comment by ymajoros** [comment]
I created a hibernate bug (though I'm not 100% sure it's not originating from Quarkus integration yet): https://hibernate.atlassian.net/browse/HHH-17750

**Comment by ymajoros** [comment]
Stack trace when GreetingResource:37 is uncommented in the reproducer:
```ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (executor-thread-1) HTTP Request to /hello/test failed, error id: fc1f7cce-b8db-4a4d-9198-d998e5e99443-1: org.hibernate.PropertyValueException: Detached entity with generated id '1' has an uninitialized version value 'null' : org.acme.entities.EntityA.version
	at org.hibernate.persister.entity.AbstractEntityPersister.isTransient(AbstractEntityPersister.java:4168)
	at org.hibernate.engine.internal.ForeignKeys.isTransient(ForeignKeys.java:293)
	at org.hibernate.type.EntityType.replace(EntityType.java:315)
	at org.hibernate.type.TypeHelper.replace(TypeHelper.java:88)
	at org.hibernate.event.internal.DefaultMergeEventListener.copyValues(DefaultMergeEventListener.java:576)
	at org.hibernate.event.internal.DefaultMergeEventListener.entityIsDetached(DefaultMergeEventListener.java:427)
	at org.hibernate.event.internal.DefaultMergeEventListener.merge(DefaultMergeEventListener.java:208)
	at org.hibernate.event.internal.DefaultMergeEventListener.doMerge(DefaultMergeEventListener.java:146)
	at org.hibernate.event.internal.DefaultMergeEventListener.onMerge(DefaultMergeEventListener.java:126)
	at org.hibernate.event.internal.DefaultMergeEventListener.onMerge(DefaultMergeEventListener.java:84)
	at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:127)
	at org.hibernate.internal.SessionImpl.fireMerge(SessionImpl.java:847)
	at org.hibernate.internal.SessionImpl.merge(SessionImpl.java:833)
	at io.quarkus.hibernate.orm.runtime.session.TransactionScopedSession.merge(TransactionScopedSession.java:156)
	at org.hibernate.engine.spi.SessionLazyDelegator.merge(SessionLazyDelegator.java:271)
	at org.hibernate.Session_OpdLahisOZ9nWRPXMsEFQmQU03A_Synthetic_ClientProxy.merge(Unknown Source)
	at org.acme.services.ServiceB.doSomething(ServiceB.java:36)
	at org.acme.services.ServiceB_Subclass.doSomething$$superforward(Unknown Source)
	at org.acme.services.ServiceB_Subclass$$function$$4.apply(Unknown Source)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:73)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:62)
	at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.invokeInOurTx(TransactionalInterceptorBase.java:136)
	at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.invokeInOurTx(TransactionalInterceptorBase.java:107)
	at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorRequired.doIntercept(TransactionalInterceptorRequired.java:38)
	at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.intercept(TransactionalInterceptorBase.java:61)
	at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorRequired.intercept(TransactionalInterceptorRequired.java:32)
	at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorRequired_Bean.intercept(Unknown Source)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:42)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:30)
	at io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:27)
	at org.acme.services.ServiceB_Subclass.doSomething(Unknown Source)
	at org.acme.services.ServiceB_ClientProxy.doSomething(Unknown Source)
	at org.acme.GreetingResource.hello(GreetingResource.java:37)
	at org.acme.GreetingResource$quarkusrestinvoker$hello_e747664148511e1e5212d3e0f4b40d45c56ab8a1.invoke(Unknown Source)
	at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)
	at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:141)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:582)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:1583)```

**Comment by cghislai** [comment]
Note: We noticed this error only occurs when the entity behind the lazy relation has a `@Version` field. I edited original post accordingly. The reproducer will be updated as well.

**Comment by marko-bekhta** [comment]
>  I created a hibernate bug (though I'm not 100% sure it's not originating from Quarkus integration yet): https://hibernate.atlassian.net/browse/HHH-17750

Thanks for taking a closer look. I have reproduced a similar error with just ORM; let's track the progress of this in the issue you've created in ORM JIRA.

**Comment by ymajoros** [comment]
@marko-bekhta I think this issue is typically mitigated in various frameworks:
- Spring has "Open session in view".
- Wildfly seems to have a similar mechanism by default (our application is migrating from Wildfly and the problem wasn't present).
- Eclipselink (JPA RI) also allows to access lazy properties of detached entities by default (not even sure you can disable that).
- `enable_lazy_load_no_trans` is mentioned in multiple sources, although I tried to activate it with `quarkus.hibernate-orm.unsupported-properties."hibernate.enable_lazy_load_no_trans"=true`, which is actually being read, but I couldn't find any usage besides configuration in Hibernate 6.4. 

So I'm wondering: shouldn't Quarkus provide a similar mechanism?

**Comment by yrodiere** [comment]
@ymajoros Feel free to open a feature request (or, more likely, link to an existing one), we can discuss that there.

Though I need to mention that "Open session in view" is [widely considered as an anti-pattern](https://www.google.com/search?client=firefox-b-d&q=%22Open+session+in+view%22+anti-pattern), so the discussion (and eventual solution) will have to address that.

**Comment by ymajoros** [comment]
Feature request: https://github.com/quarkusio/quarkus/issues/38916

**Comment by yrodiere** [comment]
Should be fixed by #40230
